### PR TITLE
feat: Added Branches tree view

### PR DESCRIPTION
**Pull request by FractalBoy**
this adds a new tree view that gives you a nice listing of all the files that changed between the current branch and the branch it was copied from.

image

tooltip:

image

clicking something marked M will open up a diff:

image

### CODE REVIEW

File: src/diffParser.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        explicitRoot: false,
+        explicitArray: false,
+        attrNameProcessors: [camelcase],
+        tagNameProcessors: [camelcase]
+      },
+      (err, result) => {
+        if (
+          err ||
+          typeof result.paths === "undefined" ||
+          !Array.isArray(result.paths.path)
```
[the remark this answers is not in the document]
Comment (FractalBoy) [reply src/diffParser.ts]: fixed in 5d852015af6e1c134f6d25ef6e6c702b61268455

**Comment by JohnstonCode** [comment]
Sorry to be a pain but can you get it to refresh when changes are detected like the repo log does. https://github.com/JohnstonCode/svn-scm/blob/master/src/historyView/repoLogProvider.ts#L122

File: src/historyView/branchChangesProvider.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+export class BranchChangesProvider
+  implements TreeDataProvider<ISvnPathChange>, Disposable {
+  private _dispose: Disposable[] = [];
+
+  dispose() {
+    dispose(this._dispose);
+  }
+
+  onDidChangeTreeData?: Event<ISvnPathChange> | undefined;
```
Comment (JohnstonCode) [inline src/historyView/branchChangesProvider.ts]: Is this used by anything?

File: src/historyView/branchChangesProvider.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        repo,
+        element.oldPath,
+        element.oldRevision,
+        element.newRevision,
+        element.newPath
+      );
+    }
+  }
+
+  constructor(private model: Model) {
```
Comment (JohnstonCode) [inline src/historyView/branchChangesProvider.ts]: Can this be moved to the top of the class after the properties

**Comment by FractalBoy** [comment]
the one feature i think maybe should be added before this is merged is a refresh button. do you agree? are there any other features that you think should be added?

**Comment by JohnstonCode** [comment]
Yes a refresh button would be useful. Apart from that i can think of anything else.

File: src/historyView/branchChangesProvider.ts (hunk 3 above)
Comment (FractalBoy) [reply src/historyView/branchChangesProvider.ts]: 169ce5177e39b0313af6a46eb82265c4588ad464

File: src/historyView/branchChangesProvider.ts (hunk 2 above)
Comment (FractalBoy) [reply src/historyView/branchChangesProvider.ts]: 0a71c09f5eca2e712f3c2fa766ededa5e7d38bc9

**Comment by FractalBoy** [comment]
> Sorry to be a pain but can you get it to refresh when changes are detected like the repo log does. https://github.com/JohnstonCode/svn-scm/blob/master/src/historyView/repoLogProvider.ts#L122

Not sure if I did this right...I don't know how to trigger the repo change event.

**Comment by JohnstonCode** [comment]
Yeah seems there is a loop. Let me have a look

**Comment by JohnstonCode** [comment]
the refresh in `getChildren` will just casue an infinate loop as its called by vscode

**Comment by FractalBoy** [comment]
that's why i put in the "fire" boolean - before it was triggering refresh, then triggering getChildren, etc.

i don't think there's an infinite loop anymore - i just don't see the event being triggered at all when i set a breakpoint

**Comment by JohnstonCode** [comment]
Yeah sorry, the `onDidChangeRepository` fires once then it just keeps refreshing

**Comment by JohnstonCode** [comment]
Got it `refresh -> repo.getChanges -> run -> updateModelState -> _onDidChangeStatus.fire -> this.model.onDidChangeRepository` then it just repeats

**Comment by FractalBoy** [comment]
hmmm...any suggestions?

**Comment by JohnstonCode** [comment]
Yeah give me a couple of minutes i will give you a patch

**Comment by JohnstonCode** [comment]
Is the https://github.com/JohnstonCode/svn-scm/pull/729/files#diff-d982bcb6ae46cbef9f548f44bd09324bR224 update needed?

**Comment by FractalBoy** [comment]
> Is the https://github.com/JohnstonCode/svn-scm/pull/729/files#diff-d982bcb6ae46cbef9f548f44bd09324bR224 update needed?

yeah, otherwise the repo may not be on the revision that was just committed (not sure why this is)

**Comment by JohnstonCode** [comment]
The information from svn info is taken directly from the .svn/entries file, and not from the server. This stems from Subversion's basic design of not requesting information from the repository server whenever possible.

**Comment by FractalBoy** [comment]
so maybe i need to do svn info on the URL instead of on the WC?

**Comment by JohnstonCode** [comment]
That would work. Also you need to add `Operation.Changes` to `isReadOnly` as its only doing log under the hood

**Comment by JohnstonCode** [comment]
Actually it wont be because of the update...

**Comment by JohnstonCode** [comment]
The manual refresh button might do for the time being. This is a complicated problem

**Comment by FractalBoy** [comment]
i'm not seeing an infinite loop - i set a breakpoint in refresh, and it's not hitting unless i click refresh

**Comment by FractalBoy** [comment]
it refreshes twice, which is maybe not optimal, but not infinite

**Comment by FractalBoy** [comment]
i think i figured it out...give me a second

**Comment by FractalBoy** [comment]
the loop is fixed, but i still don't see anything being triggered automatically when i commit a change, e.g.

**Comment by JohnstonCode** [comment]
Can you revert your last commit?

**Comment by FractalBoy** [comment]
why?

**Comment by JohnstonCode** [comment]
Think i might have it working

**Comment by FractalBoy** [comment]
alright...

**Comment by JohnstonCode** [comment]
That fixed my loop issue but the diffs arnt correct

**Comment by FractalBoy** [comment]
can i un-revert?

**Comment by FractalBoy** [comment]
(i think perhaps i'm not getting any events is because i'm on WSL)

**Comment by JohnstonCode** [comment]
Yes

**Comment by JohnstonCode** [comment]
Yes the loop is not more. The diff still isent working though. I Can see the cat is called to the corect path and with ghe correct revisions but the diff liik life its the left side against the left side

**Comment by FractalBoy** [comment]
hmm, i'm not seeing that

**Comment by JohnstonCode** [comment]
The revision on the branch if off by one it seems

**Comment by JohnstonCode** [comment]
> The revision on the branch if off by one it seems

Correct that it is just using the last revision on master and the new branch commit revision on the branch

**Comment by JohnstonCode** [comment]
it is working just not updating on future commits

**Comment by FractalBoy** [comment]
yeah, i'm seeing that too.. not sure why that's happening since i'm getting the info from the URL

**Comment by JohnstonCode** [comment]
so to fix it i need to do `svn up` then restart code then it picks up the changes

**Comment by FractalBoy** [comment]
i just need to stop relying on the cache

**Comment by FractalBoy** [comment]
also passing the url wasn't working

**Comment by FractalBoy** [comment]
should work now

**Comment by JohnstonCode** [comment]
Yes does appear to be working. Will have a proper look at it tomorrow. Rally good job though.

**Comment by FractalBoy** [comment]
thanks!

**Comment by FractalBoy** [comment]
oh, i think it's not working entirely for me since there's no .svn in the current folder. i need to modify the watcher to look at a parent path (if that's even possible)

**Comment by FractalBoy** [comment]
> Note that only files within the current workspace folders can be watched.

I guess i'll just have to deal with manually reloading since my directories are often laid out so the .svn directory is in a higher level directory

**Comment by FractalBoy** [comment]
> > Note that only files within the current workspace folders can be watched.
> 
> I guess i'll just have to deal with manually reloading since my directories are often laid out so the .svn directory is in a higher level directory

I made some changes to use `fs.watch` if the WC root is outside the workspace since `workspace.createFileSystemWatcher` only monitors files in the workspace.

**Comment by JohnstonCode** [comment]
If i open a child directory it breaks.
image

when in reality only one file has changed
image

**Comment by FractalBoy** [comment]
> If i open a child directory it breaks.
> image
> 
> when in reality only one file has changed
> image

should be fixed

**Comment by JohnstonCode** [comment]
Yeah seems to be working.

**Comment by FractalBoy** [comment]
Woohoo! Thanks!

**Pull request merged**